runtime/parachains/src/scheduler_parathreads/mod.rs

```diff
@@ -0,0 +1,159 @@
+//! - It should be possible to know at least a block ahead-of-time, ideally more,
+//!   which validators are going to be assigned to which parachains.
+//! - Parachains that have a candidate pending availability in this fork of the chain
+//!   should not be assigned.
+//! - Validator assignments should not be gameable. Malicious cartels should not be able to
+//!   manipulate the scheduler to assign themselves as desired.
+//! - High or close to optimal throughput of parachains and parathreads. Work among validator groups should be balanced.
+//!
+//! The Scheduler manages resource allocation using the concept of "Availability Cores".
+//! There will be one availability core for each parachain, and a fixed number of cores
+//! used for multiplexing parathreads. Validators will be partitioned into groups, with the same
+//! number of groups as availability cores. Validator groups will be assigned to different availability cores
+//! over time.
+
+use frame_support::pallet_prelude::*;
+use primitives::{CoreIndex, CoreOccupied, Id as ParaId, ParathreadClaim};
+use scale_info::TypeInfo;
+use sp_std::prelude::*;
+
+use crate::{
+	configuration,
+	//initializer::SessionChangeNotification,
+	paras,
+	scheduler_common::{Assignment, AssignmentProvider},
+};
+
+pub use pallet::*;
+
+//#[cfg(test)]
+//mod tests;
+
+#[derive(Encode, Decode, TypeInfo)]
+#[cfg_attr(test, derive(PartialEq, Debug))]
+/// Bounded claim queue. Bounded by config.n_parathread_cores * config.n_lookahead
+pub struct ClaimQueue {
+	queue: Vec<Vec<ParathreadClaim>>,
+	current_idx: u32,
+	bound: u32, // Is lookahead + 1
+}
+
+impl Default for ClaimQueue {
+	fn default() -> Self {
+		let queue = Vec::new();
+		Self { queue, current_idx: 0, bound: 0 }
+	}
+}
+
+impl ClaimQueue {
+	fn update(&mut self, n_lookahead: u32) {
+		self.bound = n_lookahead + 1;
+	}
+
+	fn add_claim(&mut self, claim: ParathreadClaim) {
+		match self.queue.get_mut(self.current_idx as usize) {
+			None => {
+				self.queue[self.current_idx as usize] = vec![claim];
+			},
+			Some(v) => v.push(claim),
+		}
+		self.current_idx = (self.current_idx + 1) % self.bound;
+	}
+
+	fn pop_on_idx(&mut self, idx: u32) -> Option<ParathreadClaim> {
+		self.queue.get_mut(idx as usize).and_then(|v| v.pop())
+	}
+}
+
+pub trait ClaimQueueI<T: crate::scheduler::pallet::Config> {
+	fn add_claim(&mut self, claim: ParathreadClaim) {
+		<self::Pallet<T>>::add_claim(claim)
+	}
+}
+
+impl<T: crate::scheduler::pallet::Config> AssignmentProvider<T> for Pallet<T> {
+	fn on_new_session(n_lookahead: u32) {
+		<self::Pallet<T>>::on_new_session(n_lookahead)
+	}
+
+	fn session_core_count() -> u32 {
+		<self::Pallet<T>>::session_core_count()
+	}
+
+	fn pop_assignment_for_core(core_idx: CoreIndex) -> Option<Assignment> {
+		<self::Pallet<T>>::pop_assignment_for_core(core_idx.0).map(Assignment::ParathreadA)
+	}
+
+	fn core_para(_core_idx: CoreIndex, core_occupied: &CoreOccupied) -> ParaId {
```

Comment:
ok, whether or not a core is occupied should not be part of the `AssignmentProvider` interface, this should really only live on the scheduler. AssignmentProvider just provides assignments, it does not even know that a core can be occupied/free/.... .

Reply:
This (as most of the recent changes here) were done in order to be able to test parathreads during W&S.